lib/ts/recipe/thirdparty/recipeImplementation.ts

```diff
@@ -0,0 +1,262 @@
+
+            return {
+                status: "OK",
+                url: jsonBody.url,
+                fetchResponse,
+            };
+        },
+
+        signInAndUp: async function (input: {
+            thirdPartyProviderId: string;
+            thirdPartyProviderClientId?: string;
+            thirdPartyRedirectionURL: string;
+            config: NormalisedInputType;
+            userContext: any;
+            options?: RecipeFunctionOptions;
+        }): Promise<
+            | {
+                  status: "OK";
+                  user: UserType;
+                  createdNewUser: boolean;
+                  fetchResponse: Response;
+              }
+            | {
+                  status: "NO_EMAIL_GIVEN_BY_PROVIDER";
+                  fetchResponse: Response;
+              }
+            | {
+                  status: "FIELD_ERROR";
+                  error: string;
+                  fetchResponse: Response;
+              }
+        > {
+            const stateFromStorage = this.getOAuthState({
+                userContext: input.userContext,
+                config: input.config,
+            }).state;
+
+            const code = getQueryParams("code");
+
+            const stateFromQueryParams = getQueryParams("state");
+
+            if (
+                getQueryParams("error") !== null ||
+                stateFromStorage === undefined ||
+                stateFromStorage.thirdPartyId !== input.thirdPartyProviderId ||
+                stateFromStorage.state !== stateFromQueryParams ||
+                code === null
+            ) {
+                throw new STGeneralError("");
+            }
+
+            const { jsonBody, fetchResponse } = await querier.post<
+                | {
+                      status: "OK";
+                      createdNewUser: boolean;
+                      user: UserType;
+                  }
+                | {
+                      status: "NO_EMAIL_GIVEN_BY_PROVIDER";
+                  }
+                | {
+                      status: "FIELD_ERROR";
+                      error: string;
+                  }
+            >(
+                "/signinup",
+                {
+                    body: JSON.stringify({
+                        code,
+                        thirdPartyId: input.thirdPartyProviderId,
+                        redirectURI: input.thirdPartyRedirectionURL,
+                        clientId: input.thirdPartyProviderClientId,
+                    }),
+                },
+                Querier.preparePreAPIHook({
+                    config: input.config,
+                    action: "THIRD_PARTY_SIGN_IN_UP",
+                    options: input.options,
+                    userContext: input.userContext,
+                }),
+                Querier.preparePostAPIHook({
+                    config: input.config,
+                    action: "THIRD_PARTY_SIGN_IN_UP",
+                    userContext: input.userContext,
+                })
+            );
+
```

Comment:
If the status is FIELD_ERROR, throw a STGeneralError with the message

Reply:
Done